Write a program that converts the original dataset (http://archive.ics.uci.edu/ml/machine- learning-databases/mushroom/) to a format that fits our NN’s input form. Your program should transform original dataset’s records (filename is agaricus-lepiota.data, provided in the project sub-folder, where each line represents one record with 23 features—see agaricus-lepiota.names for details) into the so-called one-hot binary code form (the first 2 numbers are classes — poisonous or edible, while others represent the remaining features—each feature may have multiple values, and you should make sure that your binary code can represent each of the values). Please take a look at data_representation_example.txt for some details. Then split the converted dataset into 3 files: training.txt for network training, val.txt for validation, and testing.txt for network testing. (You may name these files by yourself, but make sure to modify the code to fit your files). Sample_testing.txt is a sample file showing the exact format that we need.

### Import Libraries

In [12]:
import pandas as pd
import numpy as np

### Data

In [3]:
# Reading Data from the given csv file 
data = pd.read_csv('agaricus-lepiota.data', names = ['classes', 'cap-shape', 'cap-surface', 'cap-color', 'bruises?', 'odor', 'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color', 'stalk-shape', 'stalk-root', 'stalk-surface-above-ring', 'stalk-surface-below-ring', 'stalk-color-above-ring', 'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number', 'ring-type', 'spore-print-color', 'population', 'habitat'], na_values = "?")
data

classes cap-shape cap-surface cap-color bruises? odor gill-attachment  \
0          p         x           s         n        t    p               f   
1          e         x           s         y        t    a               f   
2          e         b           s         w        t    l               f   
3          p         x           y         w        t    p               f   
4          e         x           s         g        f    n               f   
...      ...       ...         ...       ...      ...  ...             ...   
8119       e         k           s         n        f    n               a   
8120       e         x           s         n        f    n               a   
8121       e         f           s         n        f    n               a   
8122       p         k           y         n        f    y               f   
8123       e         x           s         n        f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

### Exploratory Data Analysis

In [5]:
#description of the data in the DataFrame
data.describe().T

count unique top  freq
classes                   8124      2   e  4208
cap-shape                 8124      6   x  3656
cap-surface               8124      4   y  3244
cap-color                 8124     10   n  2284
bruises?                  8124      2   f  4748
odor                      8124      9   n  3528
gill-attachment           8124      2   f  7914
gill-spacing              8124      2   c  6812
gill-size                 8124      2   b  5612
gill-color                8124     12   b  1728
stalk-shape               8124      2   t  4608
stalk-root                5644      4   b  3776
stalk-surface-above-ring  8124      4   s  5176
stalk-surface-below-ring  8124      4   s  4936
stalk-color-above-ring    8124      9   w  4464
stalk-color-below-ring    8124      9   w  4384
veil-type                 8124      1   p  8124
veil-color                8124      4   w  7924
ring-number               8124      3   o  7488
ring-type                 8124      5   p  3968
spore-print-color         8124      9   w  2388
population                8124      6   v  4040
habitat                   8124      7   d  3148

### Checking for null values

In [4]:
data.isnull().sum()

classes                        0
cap-shape                      0
cap-surface                    0
cap-color                      0
bruises?                       0
odor                           0
gill-attachment                0
gill-spacing                   0
gill-size                      0
gill-color                     0
stalk-shape                    0
stalk-root                  2480
stalk-surface-above-ring       0
stalk-surface-below-ring       0
stalk-color-above-ring         0
stalk-color-below-ring         0
veil-type                      0
veil-color                     0
ring-number                    0
ring-type                      0
spore-print-color              0
population                     0
habitat                        0
dtype: int64

#### There are 2480 NaN values present in stalk-root feature it is approximately 30%. Exluding these features altogther will result in decrease in accuracy. Hence handling such values by replacing them with most frequnet value in stalk-root

In [6]:
#checking for which value of stalk-root is most frequent
data['stalk-root'].value_counts()

b    3776
e    1120
c     556
r     192
Name: stalk-root, dtype: int64

In [7]:
#relacing NaN/ ? values with 'b'
data['stalk-root'] = data['stalk-root'].fillna('b')
data['stalk-root'].value_counts()

b    6256
e    1120
c     556
r     192
Name: stalk-root, dtype: int64

In [8]:
#checking again for any null values
data.isnull().sum()

classes                     0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises?                    0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

In [9]:
# Renaming classes p and e as 0 and 1 respectively
data['classes'] =data['classes'].replace('p','0')
data['classes'] =data['classes'].replace('e','1')

In [10]:
# Transforming the categorical variable into a set of binary variables
data = pd.get_dummies(data)

In [11]:
data

classes_0  classes_1  cap-shape_b  cap-shape_c  cap-shape_f  \
0             1          0            0            0            0   
1             0          1            0            0            0   
2             0          1            1            0            0   
3             1          0            0            0            0   
4             0          1            0            0            0   
...         ...        ...          ...          ...          ...   
8119          0          1            0            0            0   
8120          0          1            0            0            0   
8121          0          1            0            0            1   
8122          1          0            0            0            0   
8123          0          1            0            0            0   

      cap-shape_k  cap-shape_s  cap-shape_x  cap-surface_f  cap-surface_g  \
0               0            0            1              0              0   
1               0            0            1              0              0   
2               0            0            0              0              0   
3               0            0            1              0              0   
4               0            0            1              0              0   
...           ...          ...          ...            ...            ...   
8119            1            0            0              0              0   
8120            0            0            1              0              0   
8121            0            0            0              0              0   
8122            1            0            0              0              0   
8123            0            0            1              0              0   

      ...  population_s  population_v  population_y  habitat_d  habitat_g  \
0     ...             1             0             0          0          0   
1     ...             0             0             0          0          1   
2     ...             0             0             0          0          0   
3     ...             1             0             0          0          0   
4     ...             0             0             0          0          1   
...   ...           ...           ...           ...        ...        ...   
8119  ...             0             0             0          0          0   
8120  ...             0             1             0          0          0   
8121  ...             0             0             0          0          0   
8122  ...             0             1             0          0          0   
8123  ...             0             0             0          0          0   

      habitat_l  habitat_m  habitat_p  habitat_u  habitat_w  
0             0          0          0          1          0  
1             0          0          0          0          0  
2             0          1          0          0          0  
3             0          0          0          1          0  
4             0          0          0          0          0  
...         ...        ...        ...        ...        ...  
8119          1          0          0          0          0  
8120          1          0          0          0          0  
8121          1          0          0          0          0  
8122          1          0          0          0          0  
8123          1          0          0          0          0  

[8124 rows x 118 columns]

### Train-Test-Val split into 90:10:10 ratio

In [13]:
train, validate, test = np.split(data.sample(frac=1, random_state= 42), [int(.8*len(data)), int(.9*len(data))])


### Saving train, test , validaion data into files

In [20]:
train.to_csv('training.txt',index= False, header=False ,encoding='utf-8')
validate.to_csv('val.txt',index= False, header=False, encoding='utf-8')
test.to_csv('testing.txt', index= False,header=False, encoding='utf-8')
